pygls/workspace.py

```diff
@@ -97,16 +91,60 @@ def _apply_incremental_change(self, change: TextDocumentContentChangeEvent) -> N
         self._source = new.getvalue()
 
     def _apply_full_change(self, change: TextDocumentContentChangeEvent) -> None:
-        """Apply a full text change to the document."""
+        """Apply a FULL text change to the document."""
         self._source = change.text
 
+    def _apply_none_change(self, change: TextDocumentContentChangeEvent) -> None:
+        """Apply a NONE text change to the document
+
+        Currently does nothing, provided for consistency.
+        """
+
     def apply_change(self, change: TextDocumentContentChangeEvent) -> None:
-        """Apply change to a document, depending on synchronization kind."""
-        try:
-            self._sync_func(change)
-        except TypeError:
-            # Don't sync document if sync kind is TextDocumentSyncKind.NONE
-            pass
+        """Apply a text change to a document, considering TextDocumentSyncKind
+
+        Performs either INCREMENTAL, FULL, or NONE synchronization based on
+        both the Client request and server capabilities.
+
+        INCREMENTAL versus FULL synchronization:
+            Even if a server accepts INCREMENTAL SyncKinds, clients may request
+            a FULL SyncKind. In LSP 3.x, clients make this request by omitting
+            both Range and RangeLength from their request. Consequently, the
+            attributes "range" and "range_length" will be missing from FULL
+            content update client requests in the pygls Python library.
+
+        Improvements:
+            Consider revising our treatment of TextDocumentContentChangeEvent,
+            and all other LSP primitive types, to set "Optional" interface
+            attributes from the client to "None". The "hasattr" check is
+            admittedly quite ugly; while it is appropriate given our current
+            state, there are plenty of improvements to be made. A good place to
+            start: require more rigorous de-serialization efforts when reading
+            client requests in protocol.py.
+        """
+        if (
+            hasattr(change, 'range') and
+            hasattr(change, 'range_length') and
```

Comment:
Attribute names passed from the client are in camelCase format.

`range_length` -> `rangeLength`